### PR TITLE
PBTS: system model made more precise

### PR DESCRIPTION
**Pull request by cason**
System model updated from comments received in meetings, being also aligned with the restrictions imposed by the formal specification of the solution in TLA+.

### CODE REVIEW

**Review by josef-widder — APPROVED** [review]
Thanks! Looks great. Even the inequalities have been updated now ;-)

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 1]
```diff
@@ -79,116 +81,163 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
 #### **[PBTS-INV-TIMELY.0]**
 
-[Time-Validity] If a correct process decides on value `v`,
-then the associated proposal time `v.time` satisfies a predefined `timely` predicate.
+- [Time-Validity] If a correct process decides on value `v`, then the proposal
+  time `v.time` was considered `timely` by at least one correct process.
+
+PBTS introduces a `timely` predicate that restricts the allowed decisions based
+on the proposal time `v.time` associated with a proposed value `v`.
+As as synchronous predicate, the time at which it is evaluated impacts on
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:141]:
```suggestion
As a synchronous predicate, the time at which it is evaluated impacts on
```

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 2]
```diff
@@ -79,116 +81,163 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+whether a process accepts of reject a value based on its proposal time.
+For this reason, the Time-Validity property refers to the previous evaluation
+of the `timely` predicate, detailed in the following.
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:144]:
```suggestion
of the `timely` predicate, detailed in the following section.
```

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 3]
```diff
@@ -79,116 +81,163 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+> The produced proposal thus have `v.round = round_p` and it is broadcast in a
+> Proposal message of with the `validRound` field set to `-1`.
+
+#### **[PBTS-PROPOSAL-RECEPTION.0]**
 
 The `timely` predicate is evaluated when a process receives a proposal.
-Let `now_p` be time a process `p` reads from its local clock when `p` receives a proposal.
-Let `v` be the proposed value and `v.time` the proposal time.
-The proposal is considered `timely` by `p` if:
+More precisely, let `p` be a correct process:
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:166]: Should this section mention `vr = -1` for this logic to be evaluated?

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 4]
```diff
@@ -79,116 +81,163 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+The existence of `POL(v,r)` is a requirement for the decision of `v` at round
+`r` of consensus.
+
+At the same time, the Time-Validity property established that if `v` is decided
+then a timely proof-of-lock `POL(v,v.round)` must have been produced.
+
+So, we need to demonstrate here that any valid `POL(v,r)` is either a timely
+proof-of-lock or it is derived from a timely proof-of-lock:
+
+#### **[PBTS-DERIVED-POL.0]**
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:223]: Really like this section! Very clear.

**Review by williambanfield — APPROVED** [review]

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md (hunk 3 above)
Comment (cason) [reply spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:166]: In the TLA+ specification we save the proposal receive time for all Proposals, in followed the same approach here. 

But in fact this time is only relevant when the proposal round matches the proposal's message round, which is the same as saying that valid round = -1.

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 5]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+PBTS is both safe and live.
+
+In addition to the variables and system parameters already defined, we use
+`beginRound(p,r)` to represent the time the process `p` reads from its clock
+when it starts round `r` of consensus.
+
+### Safety
+
+The safety of PBTS requires that if a value `v` is decided, then at least one
+correct process `p` considered the associated proposal time `v.time` timely.
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:286]: Pardon my ignorance, by why is it not that`2f+1` or even `f+1` processes have to consider it timely?

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 6]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
-we assume that the `MSGDELAY` parameter can be increased as rounds do not
-succeed on deciding a value, possibly because no proposal is considered
-`timely` by enough processes.
-The precise behavior for this workaround is under [discussion](https://github.com/tendermint/spec/issues/371).
+- the proposer of a round `r` of consensus is correct
+- and it proposes a value `v` for the first time, with associated proposal time `v.time`
+
+then the proposal `(v, v.time, r)` is accepted by every correct process provided that:
+
+- `min{p is correct : beginRound(p,r)} <= v.time <= max{p is correct : beginRound(p,r)}` and
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:328]: Is `beginRound` here the 'global' time or the local time?
From the above definition:
```
to represent the time the process `p` reads from its clock
when it starts round `r` of consensus.
```

I'm a bit unclear if it is the global time _at which_ it reads its clock or the value that it reads from its clock.


It I'm understanding correctly, this is basically saying that `v.time` must be between the times at which the processes enter the round.

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 7]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+- `max{p is correct : beginRound(p,r)} <= v.time + MSGDELAY + PRECISION <= min{p is correct : beginRound(p,r+1)}`
+
+The first condition establishes a range of safe proposal times `v.time` for round `r`.
+It is trivially observed if a correct proposer `p` sets `v.time` to the time it
+reads from its clock when starting round `r` and proposing `v`.
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:333]:
```suggestion
This condition is trivially observed if a correct proposer `p` sets `v.time` to the time it
reads from its clock when starting round `r` and proposing `v`.
```

Is this what this sentence is saying?

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 8]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+A `PROPOSAL` message sent by `p` at local time `v.time` should not be received
+by any correct process before its local clock reads `v.time - PRECISION`, so
+that condition 2. of [PBTS-TIMELY.0] is observed.
```
Comment (williambanfield) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:336]:
```suggestion
that condition 2 of [PBTS-TIMELY.0] is observed.
```

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md (hunk 5 above)
Comment (cason) [reply spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:286]: `f+1` processes means at least one correct process, that is the condition assumed here. 

For the sake of safety, we need a correct process to consider `v.time` timely. For the sake of liveness, we need all correct processes (`2f+1` of `f+1`) to consider `v.time` timely.

Comment (cason) [reply spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:286]: The definition of Time-Validity considers one correct process, it was redefined after some discussion we had.

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md (hunk 6 above)
Comment (cason) [reply spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:328]: `beginRound` is local time. The time a process reads from its clock is the time "displayed" by its clock.

I will consider a less ambiguous way to reference to the local time.

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md (hunk 7 above)
Comment (cason) [reply spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:333]: Yep, just reading the local clock before proposing should provide a `v.time` within the established range.

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md (hunk 6 above)
Comment (williambanfield) [reply spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:328]: Yeah, that sentence is a little bit ambiguous. Maybe, `the value of process `p`'s local clock when it starts round `r` of consensus`. The other way is likely ok, but maybe a little tough to parse.

**Review by williambanfield — APPROVED** [review]
A few minor fixups but otherwise looks good.

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 9]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
-There exists a system parameter `MSGDELAY` for end-to-end delays of messages carrying proposals,
-such for any two correct processes `p` and `q`, and any real time `t`:
+There exists a system parameter `MSGDELAY` for end-to-end delays of proposal messages,
+such for any two correct processes `p` and `q`:
 
-- If `p` sends a message `m` carrying a proposal at time `ts`,
-then if `q` receives the message and learns the proposal,
-`q` does that at time `t` such that `ts <= t <= ts + MSGDELAY`.
+- If `p` sends a proposal message `m` at real time `t` and `q` receives `m` at
+  real time `t'`, then `t <= t' <= t' + MSGDELAY`.
```
Comment (josef-widder) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:95]:
```suggestion
  real time `t'`, then `t <= t' <= t + MSGDELAY`.
```

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 10]
```diff
@@ -1,16 +1,31 @@
 # PBTS: System Model and Properties
 
+## Outline
+
+ - [System model](#system-model)
+   - [Synchronized clocks](#synchronized-clocks)
+   - [Message delays](#message-delays)
+ - [Problem Statement](#problem-statement)
+ - [Timely Proposals](#timely-proposals)
```
Comment (josef-widder) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:9]:
```suggestion
 - [Protocol Analysis - Timely Proposals](#protocol-analysis---timely-proposals)
```

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 11]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
-> Both [Validity] and [Time-Validity] must be observed even if up to `2f` validators are faulty.
+PBTS introduces a `timely` predicate that restricts the allowed decisions based
+on the proposal time `v.time` associated with a proposed value `v`.
+As a synchronous predicate, the time at which it is evaluated impacts on
+whether a process accepts or reject a proposal time.
+For this reason, the Time-Validity property refers to the previous evaluation
+of the `timely` predicate, detailed in the following section.
 
-### Timely proposals
+## Timely proposals
```
Comment (josef-widder) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:151]:
```suggestion
## Protocol Analysis - Timely proposals
```

**Review by josef-widder — APPROVED** [review]
looks good. I would just rename one section as proposed

File: spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md [hunk 12]
```diff
@@ -79,172 +86,264 @@ defining a lower bound, a *minimum time* that a correct process assigns to propo
...
+and observe a timely proof-of-lock.
+
+## Temporal analysis
+
+In this section we present invariants that need be observed for ensuring that
+PBTS is both safe and live.
+
+In addition to the variables and system parameters already defined, we use
+`beginRound(p,r)` to represent the time the process `p` reads from its clock
+when it starts round `r` of consensus.
```
Comment (josef-widder) [inline spec/consensus/proposer-based-timestamp/pbts-sysmodel_002_draft.md:281]:
```suggestion
beginRound(p,r)` as the value of process `p`'s local clock 
when it starts round `r` of consensus.
```

**Pull request merged**